### DISCUSSION TITLE
Implementing a Custom Force Clamp

### DISCUSSION THREAD

**Discussion by mgaimann** [glotzerlab/hoomd-blue#1831] [Support] [answered]
I am trying to implement a force clamp of the form

$F_i \mapsto k_1 * \tanh( k_2 * F_i)$, 

where

$F_i = \sum_j F_j$ 

is the sum of all individual force contributions stored in `hoomd.Simulation.operations.integrator.forces`.

In other words, I am trying to apply an operation to the total force sum in each time step, before the integration of velocities and positions is executed.
When calling `hoomd.Simulation.run()`, the order of actions should be: tuners, updaters, integrator, writer.
So I want to access the force after they have all been calculated (within the integrator), but before they are applied to the particles (also within the integrator).

Assuming the forces in the force vector are computed sequentially (is that the case?), I wonder whether I can introduce a `hoomd.md.force.Custom` that has access to the other computed forces. As far as I know, each custom force can only write to _its own_ force vector, for example via `self.cpu_local_force_arrays`.

I also tried to implement a `hoomd.custom.Action` wrapped in a `hoomd.update.CustomUpdater` that updates the accelerations, but this did not work since all relevant steps take place within the integrator I think.

Thank you for your help!

**Comment by joaander** [comment]
Use [`HalfStepHook`](https://hoomd-blue.readthedocs.io/en/v4.7.0/package-md.html#hoomd.md.HalfStepHook) which is called after net forces are computed and before the accelerations are used: https://github.com/glotzerlab/hoomd-blue/blob/6297fb37a14c425c39da2d9e77526c422b116887/hoomd/md/IntegratorTwoStep.cc#L98-L113

**Comment by mgaimann** [comment]
Hi @joaander, thank you for your answer, that is exactly what I am looking for!

Can I access the net forces in a derived Python `hoomd.md.HalfStepHook` object via a context manager similar to `cpu_local_force_arrays` in `hoomd.md.force.Custom`, or do I have to compose the half-step hook in C++, access and modify the net forces there, and then use Pybind11 to include it in the Python framework? 

Could you point me to a sample implementation of a half-step hook, if there is one available?

**Reply by joaander** [reply, marked as answer]
Use `cpu_local_snapshot` (or `gpu_local_snapshot`) in Python: https://hoomd-blue.readthedocs.io/en/v4.7.0/package-hoomd.html#hoomd.State.cpu_local_snapshot
Use `ParticleData` from C++: https://github.com/glotzerlab/hoomd-blue/blob/trunk-patch/hoomd/ParticleData.h

The only demonstration of `HalfStepHook` I see in a code search (https://github.com/search?q=repo%3Aglotzerlab%2Fhoomd-blue%20half_step&type=code) is the unit test: https://github.com/glotzerlab/hoomd-blue/blob/trunk-patch/hoomd/md/pytest/test_half_step_hook.py

**Reply by mgaimann** [reply]
Works like a charm, thank you! I missed that I can pass a handle to a simulation state object and do not need access via `._state` as for the custom forces.
Following the provided unit test example, accessing and manipulating the net forces works via:
```
with self.state.cpu_local_snapshot as snap:
    snap.particles.net_force[:] = ...
```
